secure-api-gateway-ig-extensions/src/main/java/com/forgerock/sapi/gateway/sign/SignPayloadUtil.java

```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright © 2020-2022 ForgeRock AS (obst@forgerock.com)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.forgerock.sapi.gateway.sign;
+
+import java.util.Collections;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+import org.forgerock.json.jose.builders.JwtBuilderFactory;
+import org.forgerock.json.jose.builders.SignedJwtBuilderImpl;
+import org.forgerock.json.jose.jws.JwsAlgorithm;
+import org.forgerock.json.jose.jws.SignedJwt;
+import org.forgerock.json.jose.jws.SigningManager;
+import org.forgerock.json.jose.jws.handlers.SigningHandler;
+import org.forgerock.json.jose.jwt.JwtClaimsSet;
+import org.forgerock.secrets.NoSuchSecretException;
+import org.forgerock.secrets.Purpose;
+import org.forgerock.secrets.SecretsProvider;
+import org.forgerock.secrets.keys.SigningKey;
+import org.forgerock.util.Reject;
+import org.forgerock.util.promise.Promise;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Util to sign objects <br/>
+ * Implementation of {@link SignUtil}
+ */
+public class SignPayloadUtil implements SignUtil {
+
+    private static final Logger logger = LoggerFactory.getLogger(SignPayloadUtil.class);
+    private final SigningManager signingManager;
+    private final Purpose<SigningKey> signingKeyPurpose;
+    private final Map<String, Object> critHeaderClaims;
```

Comment:
It would be better if this class were stateless and could be instantiated once and reused for all signing. This has performance benefits because we only need to run the constructor logic once, and also GC benefits as we are not creating as much garbage.

The critHeaderClaims currently prevent this being reused as these need to change per signing call (as they contain the iat claim which is time based)

Reply:
Now is in the Heap as `SapiJwsSigner` implementation.